---
title: "Analysis of a TCGA RNA-seq data set on Lung Adenocarcinoma"
author:
- name: Aranda D.
  affiliation:
  - &id Master Programme on Bioinformatics for Health Sciences, Universitat Pompeu Fabra, Barcelona, Spain
  email: david.aranda03@estudiant.upf.edu
- name: Gracia O.
  affiliation: *id
  email: oriol.gracia01@estudiant.upf.edu
- name: Serrano A.
  affiliation: *id
  email: alvaro.serrano03@estudiant.upf.edu
date: "`r format(Sys.time(), '%B %e, %Y')`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(ggplot2)
library(plyr)
library(sva)
library(grid)
library(geneplotter)
library(limma)

knitr::opts_chunk$set(echo = TRUE)

multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols))
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

# Introduction
The lung adenocarcinoma cancer is the most common type of lung cancer. It has a high incidence in the population and also an associated high mortality @Collisson2014. That’s because nowadays this cancer is really hard to diagnose in its early steps and usually is found too late @Giordano2018. Most of the previous studies aim to find genetic markers to be able to diagnose it in its stages. Although some genetic markers have been described and used, such as the mutation in the Kran gene, but more information is required @Inamura2018.

Here in this project, the data of the TCGA project will be analyzed with the aim to find (and reproduce) the finding of differential expressed genes that could help to detect this kind of cancer @Collisson2014.

# Quality assessment

## Data description

The data obtained for this project comes from The Cancer Genome Atlas (TCGA). This data contains lung cancer gene expression information of both patients and healthy individuals (control).

```{r}
se <- readRDS( "seLUAD.rds")
se
```

The first step is to import and check the provided data and it’s associated metadata. In summary it contains 600 Samples and 549 Clinical variables. Each  one of those samples contains information about the expression of 20115 genes.

```{r}
dim(colData(se))
dim(rowData(se))
```

From the Clinical variables, the ones that were found to be interesting and, then, were further analyzed were:
* Gender (Sex)
* Race (Ethnicity)
* History of other malignancy (if the patient had previous cases of cancer malignancy)
* Type (If the sample is normal or a tumor)
* Histologic Diagnosis (Information about the histological origin of the sample)

Also, each sample’s name from the TCGA contains important information that was also taken into account. The information extracted from the sample was:
* Tissue Source Site
* Sample Vial
* Sample Plate

```{r}
mcols(colData(se), use.names=TRUE)
```

To perform quality assessment and normalization the 'DGEList' object from [edgeR](http://bioconductor.org/packages/edgeR) R/Bioconductor package will be used.

```{r}
dge_luad <- DGEList(counts = assays(se)$counts, genes = as.data.frame(mcols(se)), group = se$type)
```

$\log_2$ CPM values of expression are calculated as an additional assay element to ease their manipulation.

```{r}
CPM <- t(t(dge_luad$counts)/(dge_luad$samples$lib.size/1e+06))
assays(se)$logCPM <- cpm(dge_luad, log = TRUE, prior.count = 0.25)
assays(se)$logCPM[1:3, 1:7]
```

The first  important step of this project is the assessment of the quality of the provided data. For this reason, the following analysis aims to find and correct possible errors that the data could have, such as batch effect, lack of homogeneity, possible confounding factors and source of noise (very low expressed genes).

## Sequencing depth

The first step of the quality assessment involves checking the sequencing depth of the samples. The aim of this step is to verify that all the sample groups were sequenced with the same depth distribution, and to detect if any sample group, according to each clinical variable of interest, was poorly sequenced (to see if any clinical variable has had an effect on the sequencing process).

```{r libsizes, echo=FALSE, out.width="900px", fig.cap="Library sizes in increasing order."}
ord <- order(dge_luad$sample$lib.size/1e6)
barplot(dge_luad$sample$lib.size[ord]/1e6, las=1, ylab="Millions of reads",
        xlab="Samples", col=c("blue", "red")[(se$type[ord] == "tumor") + 1])
legend("topleft", c("tumor", "normal"), fill=c("red", "blue"), inset=0.01)
```

As the number of samples is huge, a bar for each sample will be not useful because the bar would be too small for the human eye. For this reason, a different approach was taken. This approach consists of clustering the samples with similar sequencing depth and then showing a bar plot of the clusters. If the bar plots of each level (category) of the corresponding analyzed clinical variable has the same height in each of the clusters, that means that the sequencing was homogeneous for that level.

```{r custombarplot, echo=FALSE, out.width="1000px", fig.cap="Quantile barplot of lib sizes classified by factor"}
i <- 0
ordsliced <- list(0)
gender <- list(0)
race <- list(0)
typ <- list(0)
hstory_malignancy <- list(0)
for(num in 1:6){
  ordsliced[[num]] <- c(ord[(i*100 +1):(num*100)])
  i <- num
  g_nas <- length(colData(se)$gender[ordsliced[[num]]][colData(se)$gender[ordsliced[[num]]] == "NA"])
  gender[[num]] <- c(length(colData(se)$gender[ordsliced[[num]]][colData(se)$gender[ordsliced[[num]]] == "MALE"])-g_nas,
                     length(colData(se)$gender[ordsliced[[num]]][colData(se)$gender[ordsliced[[num]]] == "FEMALE"])-g_nas, g_nas)
  
  race_nas <- length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "NA"])
  race_tnas <- length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "[Not Available]"]) +
    length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "[Not Evaluated]"]) +
    length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "[Unknown]"]) -2*race_nas
  race[[num]] <- c(length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "AMERICAN INDIAN OR ALASKAN NATIVE"]) - race_nas,
                   length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "ASIAN"]) - race_nas,
                   length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "BLACK OR AFRICAN AMERICAN"]) - race_nas,
                   length(colData(se)$race[ordsliced[[num]]][colData(se)$race[ordsliced[[num]]] == "WHITE"]) - race_nas,
                   race_tnas)
  hist_nas <- length(colData(se)$history_other_malignancy[ordsliced[[num]]][colData(se)$history_other_malignancy[ordsliced[[num]]] == "NA"])
  hstory_malignancy[[num]] <- c(length(colData(se)$history_other_malignancy[ordsliced[[num]]][colData(se)$history_other_malignancy[ordsliced[[num]]] == "Yes"]) - hist_nas +
                                  length(colData(se)$history_other_malignancy[ordsliced[[num]]][colData(se)$history_other_malignancy[ordsliced[[num]]] == "Yes, History of Prior Malignancy"]) - hist_nas + 
                                  length(colData(se)$history_other_malignancy[ordsliced[[num]]][colData(se)$history_other_malignancy[ordsliced[[num]]] == "Yes, History of Synchronous and or Bilateral Malignancy"]) - hist_nas,
                                length(colData(se)$history_other_malignancy[ordsliced[[num]]][colData(se)$history_other_malignancy[ordsliced[[num]]] == "No"]) - hist_nas,
                                hist_nas)
  typ_na <- length(colData(se)$type[ordsliced[[num]]][colData(se)$type[ordsliced[[num]]] == "NA"])
  typ[[num]] <- c(length(colData(se)$type[ordsliced[[num]]][colData(se)$type[ordsliced[[num]]] == "normal"]) - typ_na,
                  length(colData(se)$type[ordsliced[[num]]][colData(se)$type[ordsliced[[num]]] == "tumor"]) - typ_na,
                  typ_na)
  
}

# Gender BARPLOT
g_df <- data.frame(supp=c("MALE", "FEMALE", "NA"),
                   depth=rep(c('1st', '2nd', '3rd', '4th', '5th', '6th'),each = 3),
                   freq = c(gender[[1]],gender[[2]],gender[[3]], 
                            gender[[4]], gender[[5]],gender[[6]]))
df_sorted <- arrange(g_df, depth, supp) 
df_cumsum <- ddply(df_sorted, "depth",transform, label_ypos=cumsum(freq))

p1 <- ggplot( data=df_cumsum, aes(x=depth, y=freq, fill=supp)) +
  geom_bar(stat="identity") +
  geom_text(aes(y=111-label_ypos,label=freq), vjust=1.6, color="black",
            position = position_dodge(0.9), size=3.5) +
  theme_minimal() + ggtitle("Barplot Gender depth distribution") +
  theme(plot.title = element_text(hjust = 0.5))

# race BARPLOT
r_df <- data.frame(supp = c("A. NATIVE", "ASIAN","BLACK", "WHITE", "NA"),
                   depth=rep(c('1st', '2nd', '3rd', '4th', '5th', '6th'),each = 5),
                   freq = c(race[[1]],race[[2]],race[[3]], race[[4]], race[[5]],race[[6]]))
df_sorted <- arrange(r_df, depth, supp) 
df_cumsum <- ddply(df_sorted, "depth",transform, label_ypos=cumsum(freq))

p2 <-ggplot( data=df_cumsum, aes(x=depth, y=freq, fill=supp)) +
  geom_bar(stat="identity") +
  geom_text(aes(y=111-label_ypos,label=freq), vjust=1.6, color="black",
            position = position_dodge(0.9), size=3.5) +
  theme_minimal() + ggtitle("Barplot Race depth distribution") +
  theme(plot.title = element_text(hjust = 0.5))

# history_other_malignanci BAPLOT
h_df <- data.frame(supp = c("Yes", "No", "NA"),
                   depth=rep(c('1st', '2nd', '3rd', '4th', '5th', '6th'),each = 3),
                   freq = c(hstory_malignancy[[1]],hstory_malignancy[[2]],hstory_malignancy[[3]], hstory_malignancy[[4]], hstory_malignancy[[5]],hstory_malignancy[[6]]))
df_sorted <- arrange(h_df, depth, supp) 
df_cumsum <- ddply(df_sorted, "depth",transform, label_ypos=cumsum(freq))

p3 <- ggplot( data=df_cumsum, aes(x=depth, y=freq, fill=supp)) +
  geom_bar(stat="identity") +
  geom_text(aes(y=111-label_ypos,label=freq), vjust=1.6, color="black",
            position = position_dodge(0.9), size=3.5) +
  theme_minimal() + ggtitle("Barplot history of other malignancy depth distribution") +
  theme(plot.title = element_text(hjust = 0.5))

#Type
t_df <- data.frame(supp = c("Normal", "Tumor", "NA"),
                   depth=rep(c('1st', '2nd', '3rd', '4th', '5th', '6th'),each = 3),
                   freq = c(typ[[1]],typ[[2]],typ[[3]], typ[[4]], typ[[5]],typ[[6]]))
df_sorted <- arrange(t_df, depth, supp) 
df_cumsum <- ddply(df_sorted, "depth",transform, label_ypos=cumsum(freq))

p4 <-ggplot( data=df_cumsum, aes(x=depth, y=freq, fill=supp)) +
  geom_bar(stat="identity") +
  geom_text(aes(y=111-label_ypos,label=freq), vjust=1.6, color="black",
            position = position_dodge(0.9), size=3.5) +
  theme_minimal() + ggtitle("Barplot of type depth distribution") +
  theme(plot.title = element_text(hjust = 0.5))

multiplot(p1, p2, p3, p4, cols=2)
```

This figure reveals substantial differences in sequencing depth between samples
and we may consider discarding those samples whose depth is substantially lower
than the rest. To identify who are these samples we may simply look at the
actual numbers including portion of the sample identifier that distinguishes them.

```{r}
sampledepth <- round(dge_luad$sample$lib.size / 1e6, digits=1)
names(sampledepth) <- substr(colnames(se), 6, 12)
head(sort(sampledepth))
```

In general, the plots of all the clinical variables showed more or less  a similar sequencing depth distribution. However, in the bar plot of the clinical variable “Type”, the plot looked a little bit odd as it seems that the normal samples were in general sequenced with less depth than the tumor ones. This is important to take into account as the final objective of this project is to check for differential expression. Unfortunately, those samples cannot be removed because they are necessary to check the differences in gene expression of healthy and cancer tissues.

## Distribution of expression levels by sample

The next step is to verify the expression levels among the samples. With this analysis we can check if any sample had an unusual expression profile. The best case scenario would be that all the samples showed a similar profile, meaning that all the samples were correctly collected. 
Because of the huge amount of samples, they will be analyzed in two plots, one for the tumor ones and other for the healthy or normal ones.

```{r distRawExp, echo=FALSE, fig.height=4, fig.width=10, out.width="1000px", fig.cap="Non-parametric density distribution of expression profiles per sample."}
par(mfrow=c(1, 2))
multidensity(as.list(as.data.frame(assays(se[, se$type == "tumor"])$logCPM)),
             xlab="log 2 CPM", legend=NULL, main="Tumor samples", las=1)
multidensity(as.list(as.data.frame(assays(se[, se$type == "normal"])$logCPM)),
             xlab="log 2 CPM", legend=NULL, main="Normal samples", las=1)
```

The results of the plots are as expected since all the samples showed a similar bimodal profile, with some density of low or non expressed genes (first peak) and some density of more expressed genes (second peak). Since no sample has had an odd expression level profile, none of them has been removed.

## Distribution of expression among genes

As it has been observed in the previous analysis, some genes seemed to have really low expression values between samples. Now, it’s interesting to see if there is any gene that is systematically lowly expressed in all the samples because that would mean that it is not a relevant gene and could lead to false positives if analyzed (small differences in low expressed genes can seem statistically significant when they are probably caused by noise). 

To check the average expression of each gene across samples, the average expression of each gene was plotted in an histogram.

```{r exprdist, echo=FALSE, out.width="650px", fig.cap="Distribution of average expression level per gene."}
avgexp <- rowMeans(assays(se)$logCPM)
par(mfrow=c(1,1))
hist(avgexp, xlab = expression(log[2] * "CPM"), main = "", las = 1, col = "gray")
```

The histogram shows that there are in fact some systematically lowly expressed genes that should be removed. To filter out those genes, two cut-off values were used: the first one entailed a minimum CPM value accepted (obtained from: Minimum logCPM = 10/minimum depth). After deleting the genes that did not pass the first cut-off another filter step was applied. This filter step involved removing the genes that, because of the first cut-off step, had less samples than the smallest gene group (which in this case is the normal group, with 59 samples). 

```{r}
cpmcutoff <- round(10/min(dge_luad$sample$lib.size/1e+06), digits = 1)
cpmcutoff

nsamplescutoff <- min(table(se$type))
nsamplescutoff

mask <- rowSums(cpm(dge_luad) > cpmcutoff) >= nsamplescutoff

se.filt <- se[mask, ]
dge_luad.filt <- dge_luad[mask, ]
dim(se.filt)
```

The results of the previous filtering are represented in the same histogram plot with red bars. These are the genes that passed the whole filtering process.

At the beginning, we had 20115 different genes but after the filtering of the low expressed ones, `r dim(se.filt)[1]` remained.

```{r filtexprdist, echo=FALSE, out.width="650px", fig.cap="Distribution of average expression level of filtered and non filtered genes."}
h <- hist(avgexp, xlab = expression("Expression level (" * log[2] * "CPM)"), main = "", 
          las = 1, col = "grey", cex.axis = 1.2, cex.lab = 1.5)
x <- cut(rowMeans(assays(se.filt)$logCPM), breaks = h$breaks)
lines(h$mids, table(x), type = "h", lwd = 10, lend = 1, col = "darkred")
legend("topright", c("All genes", "Filtered genes"), fill = c("grey", "darkred"))
```

## Normalization

One vital step is the quantile normalization, which is needed because, after sequencing, we obtain a given number of total reads that is fixed by the experiment procedure.
This means that the counts observed for each gene are restricted by the total amount of reads, so if a sample has more expressed genes than another one with the same amount of reads, the numbers will not be equivalent and thus not comparable. ( Two genes equally expressed in two different samples can have different values of counts if one of those samples has more expressed genes than the other one).

```{r}
dge_luad.filt <- calcNormFactors(dge_luad.filt, normalize.method="quantile")
```

To see how this normalized step has affected to the expression levels of our data, two vuvuzella plots have been made, one with the non-filtered non-normalized data (left) and the other with the normalized and filtered one (right). As we can see, the differences are not really high which means that the data still has lots of variability.

```{r vuvuprefilt, echo=FALSE, out.width="1000px", fig.cap="Vuvuzela plots"}
par(mfrow = c(1, 2))
plotSmear(dge_luad, lowess = TRUE, las = 1, cex.lab = 1.5, cex.axis = 1.2, main = "Preprocesed data")
abline(h = 0, col = "blue", lwd = 2)
plotSmear(dge_luad.filt, lowess = TRUE, las = 1, cex.lab = 1.5, cex.axis = 1.2, main = "Filtered and normalized data")
abline(h = 0, col = "blue", lwd = 2)
```

After the normalization, the bulk of the difference of the mean expression between tumor and normal samples was closer to 0. This indicates that the process was done correctly since we expect that most of the genes are not differential expressed (so the bulk of the mean should be close to 0).

## MA-plots

The objective of this set of plots is to see if any sample as an expression profile that is different from the rest of the samples. As the gene with differential expression should be few, the majority of genes should have an expression level similar to the average expression for that gene. So, in summary, the most of the points should be around the 0 (no difference).

The profiles for all the samples had been observed. However, due to the huge amount of samples only 50 of the two important groups ( tumor and normal) are shown.

```{r maPlots, fig.height=36, fig.width=8, dpi=100, echo=FALSE, fig.cap="MA-plots of some tumor(blue) and normal(green) samples."}
par(mfrow=c(12, 4), mar = c(2,2,2,2))
for (i in c(1,553)) {
  inici <- i
  final <- i+47
  for (i in inici:final) {
    A <- rowMeans(assays(se.filt)$logCPM) ; M <- assays(se.filt)$logCPM[, i] - A; C <- se.filt$type[[i]];
    smoothScatter(A, M,
                  main=colnames(se.filt)[i],
                  las=1,
                  ylim=c(-10,10),
                  cex.axis=1.2,
                  cex.lab=1.5,
                  cex.main=1,
                  colramp = colorRampPalette(c("white",ifelse(C == "tumor", "blue","green")),space = "Lab"))
    abline(h=0, col="blue", lwd=2) ; lo <- lowess(M ~ A) ; lines(lo$x, lo$y, col="red", lwd=2)
  }
}

```

The result don’t show expression-level dependent biases among the samples, so no sample needs to be removed.

## Variability and batch identification

### Check sources of cofounding variability

To begin with, a MDS plot was done to check which is the main source of variability of our data.

```{r firstmds, echo=FALSE, out.width="1000px", fig.cap="MDS preprocessed"}
par(mfrow = c(1,1), mar = c(5.1,4.1,4.1,2.1))
plotMDS(dge_luad.filt, col = c("red", "blue")[as.integer(dge_luad.filt$samples$group)], cex = 0.7, main = "MDS plot: Normal vs tumor")
legend("topright", c("Normal", "Tumor"), fill = c("red", "blue"), inset = 0.05, cex = 1)
```

As expected, the main source of variability is the difference between tumor and normal samples.
In addition, after doing the plot, two possible outliers were detected. These are 
TCGA.64.5775.01A.01R.1628.07 and perhaps TCGA.49.AAR9.01A.21R.A41B.07 samples. 
These two samples will not be removed yet but they will be examined with detail in further analysis.

#### Identifying batch effect variables

The variables in which the batch effect will be checked are the following:

```{r}
tss <- substr(colnames(se.filt), 6, 7)
center <- substr(colnames(se.filt), 27, 28)
plate <- substr(colnames(se.filt), 22, 25)
portionanalyte <- substr(colnames(se.filt), 18, 20)
samplevial <- substr(colnames(se.filt), 14, 16)
gender <- unname(se.filt$gender)
race <- unname(se.filt$race)
histo <- unname(se.filt$histologic_diagnosis.1)
```

In order to check possible batch effects in the clinical variables of interest, cross tabulation tables were performed.

```{r}
table(data.frame(TYPE=se.filt$type, TSS=tss))
# Too much zeros. Only six TSS have noral samples

table(data.frame(TYPE=se.filt$type, TSS=samplevial))
# hey batch effect!!

table(data.frame(TYPE=se.filt$type, PLATE=plate))
# Too much zeros: only 5 plates with normal. 
# One of the plates (1758) only has normals!!!

table(data.frame(TYPE=se.filt$type, PORTIONALYTE=portionanalyte))
# All normal samples have extremly low values of portion analyte
# Not necessary: tumor and normal have separated values

table(data.frame(TYPE=se.filt$type, GENDER=gender))
#Seems pretty equilibred

table(data.frame(TYPE=se.filt$type, RACE=race))
#Mising too much dat

table(data.frame(TYPE=se.filt$type, RACE=histo))
#Mising too much dat
```

Most tables didn’t present clear results as the amount of normal samples is very low compared to the tumor ones. 
However, the sample vial variable showed clearly biased results, as all the normal samples were sequenced in different vials.

### Batch effect found in sample vial

As the sample vial variable showed a clear batch effect in the previous analysis, different batch effect removing techniques will be used to decrease the impact of this batch in our data.

```{r samplevialdendo1, fig.height=7, fig.width=14, dpi=100, echo=FALSE, fig.cap="Dendogram of variation labeled by Sample Vial"}
logCPM <- cpm(dge_luad.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(factor(samplevial))
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- as.character(se.filt$type)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples: Sample Vial")
legend("topright", legend=sort(unique(batch)), 
       fill=sort(unique(batch)), inset=0.05, cex = 0.7)
```

The figure /@ref(fig:samplevialdendo1) provides a a way to graphically observe this batch effect.

The first approach used to remove the batch effect was QR-decomposition.
```{r samplevialdendo2, fig.height=7, fig.width=14, dpi=100, echo=FALSE, fig.cap="Dendogram of variation labeled by Sample Vial corrected with QR", message=FALSE}
mod <- model.matrix(~ se.filt$type, colData(se.filt))
qrexp <- removeBatchEffect(logCPM, batch, design = mod)
d <- as.dist(1-cor(qrexp, method="spearman"))
sampleClustering <- hclust(d)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples: Sample Vial - QR correction")
legend("topright", legend=sort(unique(batch)), 
       fill=sort(unique(batch)), inset=0.05, cex = 0.7)
```

As it can be seen, this method didn't accomplish the desired effect since the batch effect is still present. Thus, a more agressive method, namely SVD batch removal, was applied:

```{r samplevialdendo3, fig.height=7, fig.width=14, dpi=100, echo=FALSE, fig.cap="Dendogram of variation labeled by Sample Vial corrected with SVD", message=FALSE}
library(corpcor)
s <- fast.svd(t(scale(t(logCPM), center = TRUE, scale = TRUE)))
pcSds <- s$d
pcSds[1] <- 0
svdexp <- s$u %*% diag(pcSds) %*% t(s$v)
colnames(svdexp) <- colnames(se.filt)
d <- as.dist(1-cor(svdexp, method="spearman"))
sampleClustering <- hclust(d)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples: Sample Vial - SVD correction")
legend("topright", legend=sort(unique(batch)), 
       fill=sort(unique(batch)), inset=0.05, cex = 0.7)
```

Unfortunately, after the statistical treatment the sample vial batch effect is still present. 
As the normal samples can’t be removed, both approaches will be applied simultaneously to try to remove the effect. 

```{r samplevialdendo4, fig.height=7, fig.width=14, dpi=100, echo=FALSE, fig.cap="Dendogram of variation labeled by Sample Vial corrected with SVD & QR", message=FALSE}
logCPM.batch <- logCPM
logCPM <-removeBatchEffect(logCPM, batch, design = mod)
assays(se.filt)$logCPM <- logCPM
s <- fast.svd(t(scale(t(logCPM), center = TRUE, scale = TRUE)))
pcSds <- s$d
pcSds[1] <- 0
svdexp <- s$u %*% diag(pcSds) %*% t(s$v)
colnames(svdexp) <- colnames(se.filt)
d <- as.dist(1-cor(svdexp, method="spearman"))
sampleClustering <- hclust(d)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples: Sample Vial - QR & SVD correction")
legend("topright", legend=sort(unique(batch)), 
       fill=sort(unique(batch)), inset=0.05, cex = 0.7)

```

After applying the combination of both methods the batch effect was finally removed. 
Now it’s important to check that this aggressive approach hasn’t removed the variability of our data. 
This topic will be covered in the next section

### Analysis of unclear variables 

There are variables where batch effect presence is uncertain. Thus MDS plot were done as a graphical way to check possible batch effects.
This unclear variables were:
  + Tisue source site (TSS)
  + Plate
  + Sex
  + Ethnicity
  + Histology
Sample vial variable, which is pressumed to be associated with batch effect, was ploted as reference: 

```{r mdsplots1, fig.height=7, fig.width=14, dpi=100, echo=FALSE, fig.cap="MDS plots without batch labeled for different clinical variables", message=FALSE}
titles_batchs <- c("MDS plot: Tisue Source Site",
                   "MDS plot: Plate",
                   "MDS plot: Sex",
                   "MDS plot: Ethnicity",
                   "MDS plot: Histology")
objects_batchs <- list(tss, samplevial, plate, gender,race, histo)
outcome <- paste(substr(colnames(se.filt), 9, 12), as.character(se.filt$type), sep="-")
for (a in 1:5){
  par(mfrow=c(1,1))
  d <- as.dist(1-cor(logCPM, method="spearman"))
  sampleClustering <- hclust(d)
  batch <- factor(objects_batchs[[a]])
  plotMDS(dge_luad.filt, labels=outcome, col=as.integer(batch), main =titles_batchs[a])
  legend("bottomright", legend=sort(unique(batch)), 
         fill=sort(unique(batch)), inset=0.05, cex = 0.7)
}
```
There doesn't seems to be any significant batch effect across these uncertain variables (except for sample vial, as was expected). Therefore, the uncertainty observed on cross tabulation tables was due to sample size differences between normal and tumor groups.

# Differential expression

Since the main objective of this project was to find differentially expressed genes an initial rough approach is used. 
However, a more complex and detailed analysis will be performed in the second part of the project.
The most basic method to spot differential expression is to check the standard deviation of expression of each gene between samples against the mean expression. The genes with a higher standard deviation than the expected from the mean expression are the best candidates to be differentially expressed.

```{r}
logCPM.filt <- logCPM
logCPM.origin <- cpm(dge_luad, log=TRUE, prior.count = 3)
meanUnloggedExp.filt <- rowMeans(assays(se.filt)$counts)
sdUnloggedExp.filt <- apply(assays(se.filt)$counts, 1, sd)
meanLoggedExp.filt <- rowMeans(logCPM.filt)
sdLoggedExp.filt <- apply(logCPM.filt, 1, sd)

meanUnloggedExp <- rowMeans(assays(se)$counts)
sdUnloggedExp <- apply(assays(se)$counts, 1, sd)
meanLoggedExp <- rowMeans(logCPM.origin)
sdLoggedExp <- apply(logCPM.origin, 1, sd)
```

```{r DElogynormal, echo=FALSE, out.width="900px", fig.cap="Scatter plot: variability vs gene expression"}
par(mfrow=c(1, 2))
plot(meanLoggedExp, sdLoggedExp, pch=".", cex=4, xlab="Log mean expression", ylab="SD", main= "Differential expression preprocessed")
lines(lowess(meanLoggedExp, sdLoggedExp, f=0.25), lwd=2, col="red")
plot(meanLoggedExp.filt, sdLoggedExp.filt, pch=".", cex=4, xlab="Log mean expression", ylab="SD", main= "Differential expression filtered")
lines(lowess(meanLoggedExp.filt, sdLoggedExp.filt, f=0.25), lwd=2, col="red")
```

To combine the data apported by both Standard Deviation and log mean expressions, multiple F-tests are performed (one for each gene). 
```{r}
mod <- model.matrix(~ se.filt$type, colData(se.filt))
mod0 <- model.matrix(~ 1, colData(se.filt))
pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
sum(p.adjust(pv, method = "bonferroni") < 0.01)
```

There are `r sum(p.adjust(pv, method="fdr") < 0.01)` genes changing significantly
their expression at FDR < 1%. In Figure \@ref(fig:pdist) below  the distribution of the
resulting p-values is showed:

```{r pdist, echo=FALSE, out.width="1000px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples."}
hist(pv, main="", las=1)
```

However, this kind of tests often give a lot of False Positives. Probably, most of this XXXX genes are not actually differentially expressed. Thus, we’ll use the surrogate variables (SVA) method from the R/Bioconductor package [sva](http://bioconductor.org/packages/sva) to reduce the number of differentially expressed genes.

```{r}
sv <- sva(assays(se.filt)$logCPM, mod, mod0)
sv$n
```

The SVA algorithm has found `r sv$n` surrogate variables. Let's use them to
assess against the extent of differential expression this time adjusting for these
surrogate variables.

```{r}
modsv <- cbind(mod, sv$sv)
mod0sv <- cbind(mod0, sv$sv)
pvsv <- f.pvalue(assays(se.filt)$logCPM, modsv, mod0sv)
sum(p.adjust(pvsv, method="bonferroni") < 0.01)
```
As was expected, the number of differentially expressed genes has been reduced significantly.

## Volcano Plot

Nonetheless, significance in the differential expression is not enough to provide this genes with a clinical interest. The fold change is the other factor to be noticed and it represents the magnitude of differential expression of each gene between the study groups (in this particular case tumor and normal).

The volcano plot encompass both features, making it a perfect tool to visually understand our data. 

```{r volcano, echo=FALSE, out.width="600px", fig.cap="Volcano plot"}
par(mfrow = c(1,1))
logFC <- tumorExp-normalExp
plot(logFC, -log10(pv), pch=16, cex=0.7, xlab="Log fold-change", ylab="-log10 Raw p-value", las=1, col = ifelse(abs(logFC) >= 2, "red", "black"))
abline(h=-log10(max(pvsv[FDRpvalue <= 0.001])), lty=2)
```

In red, the genes that are both with significant differential expression and a high enough fold change (arbitrary set to |2|).In conclussion, this genes are the best candidates to have a role in the malignancy of tumor cells or in some aspect of lung adenocarcinoma tumor.

# Session information

For reproducibility purposes we provide the session information.

```{r}
sessionInfo()
```

#####################
# Second part scratch
#####################

## Charge data again

Charge and filter data again

```{r}
library(SummarizedExperiment)
se <- readRDS( "seLUAD.rds")

#ChargeDGE list and normalization factor
library(edgeR)
dge_luad <- DGEList(counts = assays(se)$counts, group = se$type, genes = as.data.frame(rowData(se)))
assays(se)$logCPM <- cpm(dge_luad, log = TRUE, prior.count = 0.25)
dge_luad <- calcNormFactors(dge_luad)

#Filter out low-expresed genes
mask <- rowMeans(assays(se)$logCPM) > 1

se <- se[mask, ]

dge_luad <- dge_luad[mask, ]
dim(dge_luad)
```
Start limma DE analysis

```{r}
#1. Build design matrix
mod <- model.matrix(~ se.filt$type, colData(se.filt))
head(mod)

#2. Fit linear regression models
fit <- lmFit(logCPM.filt, mod)
class(fit)
names(fit)

#3. Calculate Bayes
fit <- eBayes(fit)
class(fit)
names(fit)

#4. Output results
res <- decideTests(fit)
summary(res)
```

Filter results by FDR
```{r}
FDRcutoff <- 0.05
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

Full table
```{r}
tt <- topTable(fit, coef = 2, n = Inf)
head(tt)
```

Show metadata for biological interpretation
```{r}
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 
    1], stringsAsFactors = FALSE)# We have to say explicitly metadata is not a factor
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt)
```

Now, let's examine chromosome distribution of DE genes at 10% FDR
```{r}
sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE)
```

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(-1.5, 10, lwd = 2)
```

Try again, with covariates
Repeat pipeline adding the known covariate (concentration of RNA in sample, in this case)
```{r}
#1. Build design matrix
mod <- model.matrix(~ se.filt$type, colData(se.filt))
head(mod)

v <- voom(dge_luad.filt, mod, plot=TRUE)

#2. Fit linear model
fit5 <- lmFit(v, mod)

#3. T-statistics
fit5 <- eBayes(fit5)

#4. Extend of differencial expression
res5 <- decideTests(fit5, p.value = FDRcutoff)
summary(res5)

#5. Metadata and fetch table
fit5$genes <- genesmd
tt5 <- topTable(fit5, coef = 2, n = Inf)
head(tt5)

#6. Chromosome distribution
sort(table(tt5$chr[tt5$adj.P.Val < FDRcutoff]), decreasing = TRUE)

#7. Assess accuracy
DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff]
r <- c(r, sum(DEgenes5 %in% c(msy, xie))/length(c(msy, xie)))
r

p <- c(p, sum(DEgenes5 %in% c(msy, xie))/length(DEgenes5))
p

#8. Diagnostic plots
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt5$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit5$t[, 2], df = fit5$df.prior + fit5$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

```{r}
volcanoplot(fit5, coef = 2, highlight = 7, names = fit5$genes$symbol, main = "Model 5", 
    las = 1)
```

